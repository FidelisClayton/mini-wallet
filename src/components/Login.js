import React, { Component } from 'react'
import { css } from 'emotion'
import md5 from 'md5'

import Button from './Button'
import LabeledInput from './LabeledInput'
import Card from './Card'

import * as colors from '../helpers/colors'

const styles = css({
  transition: `
    500ms ease margin-top 200ms,
    500ms ease transform
  `,

  '.login': {
    '&__error': {
      color: colors.red,
      fontWeight: 'bold'
    },
  },

  '&.login--inactive': {
    position: 'absolute',
    marginTop: '-100px',
    transform: 'scale(0.90)',
    zIndex: 0
  },

  '.card__title': {
    cursor: 'pointer'
  }
})

const initialState = {
  data: {
    email: '',
    password: '',
  },
  errors: []
}

export class Login extends Component {
  state = initialState

  get classNames () {
    const classNames = [
      'login',
      !this.props.active && 'login--inactive',
      styles
    ]
      .filter(className => !!className)
      .join(' ')

    return classNames
  }

  handleSubmit = (event) => {
    event.preventDefault()
    event.stopPropagation()

    const data = this.state.data

    this.props.onSubmit({
      email: data.email,
      password: md5(data.password)
    })
  }

  handleInputChange = (inputName) => (event) => {
    this.setState({
      data: {
        ...this.state.data,
        [inputName]: event.target.value
      }
    })
  }

  render () {
    const {
      active,
      onClick,
      error
    } = this.props

    return (
      <Card className={this.classNames}>
        <h4
          className="card__title"
          onClick={onClick}
        >
          Login
        </h4>

        <form onSubmit={this.handleSubmit}>
          <LabeledInput
            label="Email"
            inputType="email"
            onChange={this.handleInputChange('email')}
            inputProps={{
              defaultValue: this.state.data.email
            }}
          />

          <LabeledInput
            label="Senha"
            inputType="password"
            onChange={this.handleInputChange('password')}
            inputProps={{
              defaultValue: this.state.data.password
            }}
          />

          { active && error && (
            <span className="login__error">{ error }</span>
          )}

          <div className="card__submit">
            <Button>
              Entrar
            </Button>
          </div>
        </form>
      </Card>
    )
  }
}

export default Login
