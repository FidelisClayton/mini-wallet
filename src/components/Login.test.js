import React from 'react'
import { shallow } from 'enzyme'

import Login from './Login'

describe('Component: <Login />', () => {
  it('should render properly', () => {
    const wrapper = shallow(
      <Login
        onClick={jest.fn()}
        error={null}
        active={true}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render error', () => {
    const wrapper = shallow(
      <Login
        onClick={jest.fn()}
        error={'Error'}
        active={true}
      />
    )

    expect(wrapper.find('.login__error').text()).toEqual('Error')
  })

  it('should not render error when Login is not active', () => {
    const wrapper = shallow(
      <Login
        onClick={jest.fn()}
        error={'Error'}
        active={false}
      />
    )

    expect(wrapper.find('.login__error')).toHaveLength(0)
  })

  it('should handle click on title', () => {
    const onClick = jest.fn()

    const wrapper = shallow(
      <Login
        onClick={jest.fn()}
        error={null}
        active={true}
        onClick={onClick}
      />
    )

    wrapper.find('.card__title').simulate('click')

    expect(onClick.mock.calls.length).toEqual(1)
  })
})
