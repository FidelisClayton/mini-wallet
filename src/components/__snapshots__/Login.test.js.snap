// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Login /> should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    active={true}
    error={null}
    onClick={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h4
          className="card__title"
          onClick={[Function]}
>
          Login
</h4>,
        <form
          onSubmit={[Function]}
>
          <LabeledInput
                    inputProps={
                              Object {
                                        "defaultValue": "",
                                      }
                    }
                    inputType="email"
                    label="Email"
                    onChange={[Function]}
          />
          <LabeledInput
                    inputProps={
                              Object {
                                        "defaultValue": "",
                                      }
                    }
                    inputType="password"
                    label="Senha"
                    onChange={[Function]}
          />
          <div
                    className="card__submit"
          >
                    <Button>
                              Entrar
                    </Button>
          </div>
</form>,
      ],
      "className": "login css-ec5zgr",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
          "className": "card__title",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Login",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <LabeledInput
              inputProps={
                            Object {
                                          "defaultValue": "",
                                        }
              }
              inputType="email"
              label="Email"
              onChange={[Function]}
/>,
            <LabeledInput
              inputProps={
                            Object {
                                          "defaultValue": "",
                                        }
              }
              inputType="password"
              label="Senha"
              onChange={[Function]}
/>,
            null,
            <div
              className="card__submit"
>
              <Button>
                            Entrar
              </Button>
</div>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "inputProps": Object {
                "defaultValue": "",
              },
              "inputType": "email",
              "label": "Email",
              "onChange": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "inputProps": Object {
                "defaultValue": "",
              },
              "inputType": "password",
              "label": "Senha",
              "onChange": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Button>
                Entrar
</Button>,
              "className": "card__submit",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Entrar",
              },
              "ref": null,
              "rendered": "Entrar",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h4
            className="card__title"
            onClick={[Function]}
>
            Login
</h4>,
          <form
            onSubmit={[Function]}
>
            <LabeledInput
                        inputProps={
                                    Object {
                                                "defaultValue": "",
                                              }
                        }
                        inputType="email"
                        label="Email"
                        onChange={[Function]}
            />
            <LabeledInput
                        inputProps={
                                    Object {
                                                "defaultValue": "",
                                              }
                        }
                        inputType="password"
                        label="Senha"
                        onChange={[Function]}
            />
            <div
                        className="card__submit"
            >
                        <Button>
                                    Entrar
                        </Button>
            </div>
</form>,
        ],
        "className": "login css-ec5zgr",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
            "className": "card__title",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Login",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <LabeledInput
                inputProps={
                                Object {
                                                "defaultValue": "",
                                              }
                }
                inputType="email"
                label="Email"
                onChange={[Function]}
/>,
              <LabeledInput
                inputProps={
                                Object {
                                                "defaultValue": "",
                                              }
                }
                inputType="password"
                label="Senha"
                onChange={[Function]}
/>,
              null,
              <div
                className="card__submit"
>
                <Button>
                                Entrar
                </Button>
</div>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "inputProps": Object {
                  "defaultValue": "",
                },
                "inputType": "email",
                "label": "Email",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "inputProps": Object {
                  "defaultValue": "",
                },
                "inputType": "password",
                "label": "Senha",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button>
                  Entrar
</Button>,
                "className": "card__submit",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Entrar",
                },
                "ref": null,
                "rendered": "Entrar",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
